Overhaul partition handling

rolandbeisel opened the pull request.

Alek96 commented on namastack-outbox-core/src/main/kotlin/io/namastack/outbox/partition/NewInstancesRebalancer.kt:
```diff
@@ -0,0 +1,58 @@
+        if (newInstanceIds.isEmpty()) return
+
+        val partitionsToReleaseCount =
+            max(0, partitionContext.countOwnedPartitionAssignments() - partitionContext.targetPartitionCount)
```
Consider moving it to the `PartitionContext` class, e.g.,
```kotlin
fun countPartitionsToRelease(): Int = 
  max(0, countOwnedPartitionAssignments() - targetPartitionCount)

val partitionsToReleaseCount = partitionContext.countPartitionsToRelease()
```

Alek96 commented on namastack-outbox-jpa/src/main/kotlin/io/namastack/outbox/JpaOutboxPartitionAssignmentRepository.kt:
```diff
@@ -0,0 +1,177 @@
+    ) = transactionTemplate.executeNonNull {
+        val now = OffsetDateTime.now(clock)
+
+        val entities =
```
A for loop could be better here:
```kotlin
override fun claimStalePartitions(
    partitionIds: Set<Int>,
    newInstanceId: String,
) = transactionTemplate.executeNonNull {
    val now = OffsetDateTime.now(clock)

    for (partitionNumber in partitionIds) {
        val entity = entityManager.find(OutboxPartitionAssignmentEntity::class.java, partitionNumber)
            ?: throw IllegalStateException("Partition cannot be claimed: missing",)
    
        entity.reassignTo(newInstanceId, now)
        entity.let { entityManager.merge(it) }
    }
}
```

Alek96 reviewed (approved):
I’ve added a few comments, but they’re mostly related to my personal coding style. Overall, it’s a +1 from me.

rolandbeisel commented:
> I’ve added a few comments, but they’re mostly related to my personal coding style. Overall, it’s a +1 from me.

Hey @Alek96, thanks again for reviewing my code! I’ve done some overall refactoring, and I’d be really happy if you could find a moment to look over the changes. Hopefully most of the points from your last review are also addressed now.

rolandbeisel commented on namastack-outbox-core/src/main/kotlin/io/namastack/outbox/partition/PartitionAssignmentRepository.kt:
```diff
@@ -0,0 +1,77 @@
+     */
+    fun claimStalePartitions(
+        partitionIds: Set<Int>,
+        staleInstanceIds: Set<String>?,
```
I now use the (existing) version field for optimistic locking and pass in a `Set<PartitionAssignment>`. It's now very simple, which I like a lot :)
